Provide an option to not unload OpenSSL on dlclose()

mattcaswell opened the pull request:
By default OpenSSL will clean itself up via an atexit() handler. Most platforms will call the atexit() handler on dlclose() if OpenSSL has been dynamically loaded at runtime. However, some platforms only attempt to call the atexit() handler on process termination which may cause crashes if OpenSSL has been dynamically loaded. This PR adds an option (no-unload-library) to force OpenSSL to remain loaded up until process termination to avoid these crashes. It does this by calling dlopen() internally and deliberately leaking the returned reference. This leak is harmless but may be detected by mem leak checkers.

This PR also fixes an atexit() issue on Windows. The Windows implementation of atexit() does not get called on FreeLibrary() (the equivalent of dlclose()). However the MS function _onexit() does. Therefore we use that instead. This means that use of no-unload-library should not be necessary on Windows platforms.

This is an alternative approach to the same problem given in #1690 as per the suggestion in that PR.

This is marked as WIP because it still needs a test.

mattcaswell commented:
I'd like feedback on whether #1690 or this PR is preferred.

vdukhovni commented:
This needs a bit more work.  I asked Nico Williams to comment, perhaps later today.

kaduk commented:
Do we have a list of platforms that do/do not call the atexit/_onexit()-registered functions on dlclose()?
While "require the use of RTLD_NODELETE" is perhaps the most aesthetically pleasing route, I think experience has shown us that requiring actions of callers (even when well-documented) does not get anywhere close to full coverage, and it's better to make things automatically "just work" to the extent possible.

mattcaswell commented:
I made some notes on a few different platforms where I could figure out the status from online sources:

Solaris 10 (25th May 2001)

       The atexit() function registers the function pointed to by func  to  be
       called  without  arguments on normal termination of the program or when
       the object defining the function is unloaded.

       Normal termination occurs by either a call to the exit(3C) function  or
       a   return  from  main().  Object  unloading  occurs  when  a  call  to
       dlclose(3C) results in the object becoming unreferenced.


Linux (Since glibc 2.2.3 - 26th April 2001)
       Since  glibc  2.2.3,  atexit()  (and  on_exit(3))  can be used within a
       shared library to establish functions that are called when  the  shared
       library is unloaded.
       
Windows - use _onexit instead

OpenBSD
       Documented behaviour as per Linux. The documentation only aquired a note
       about shared library behaviour in November 2014. However the commit
       message suggests that the behaviour has been like that for some while but
       no-one knows when it originated.
       Looking at the source code I *think* the change was implemented in
       OpenBSD 4.3. The relevant CHANGELOG line is
       "Add __cxa_atexit() support to gcc3, provides support for shared object
       destructors called at dlclose() time."
       Available here: https://www.openbsd.org/plus43.html
       OpenBSD 4.3 was released in May 2008. The relevant commit message talks
       about this being "inspired by similar changes in FreeBSD and NetBSD."
       The commit can be seen here:
       https://github.com/bluerise/openbsd-src/commit/5198527c0e2957645fda7040dfcb34da8197dcbd

FreeBSD
       Makes no mention of this behaviour in the docs. However I found this
       patch from 2003 which adds this capability:
       https://lists.freebsd.org/pipermail/freebsd-bugs/2003-November/004146.html
       And it seems to have later been commited (Dec 2003):
       https://github.com/freebsd/freebsd/commit/caabe905470f909266101f4eae7e6651e6c58083
       
MacOS
       I don't have easy access to MacOS. The online atexit docs make no mention of
       the behaviour on DSO unload, but AFAICT the implementation seems to have been
       lifted from FreeBSD and does support the DSO unload functionality:
       http://opensource.apple.com/source/Libc/Libc-1081.1.3/stdlib/FreeBSD/atexit.c

nicowilliams commented:
I would suggest two different things:

 - if possible use .fini sections ("destructors" in GCC parlance) instead of atexit()
 - where possible use dladdr() with the address of a static variable to get the absolute path of the _current_ object, then dlopen() that

The dladdr() thing is important because the object that you might find using dlopen() otherwise might not be the same as the object calling dlopen()!  This can happen if, for example, the calling object was loaded with an absolute path and the process' run-path / LD_LIBRARY_PATH would find a different libcrypto or libssl.  It can also happen if you're using a different SONAME than the one you're hardcoding in the source code...  (ask @vdukhovni about that).

A third comment is that this should probably not be disablable.  If you get false positive memory leak reports from valgrind or what have you, well, too bad -- that's what suppressions (and the valgrind APIs) are for.

nicowilliams commented:
Also, on Windows perhaps the right thing is to use DllMain()...

mattcaswell commented:
Thanks @nicowilliams. There are a number of things I don't understand about your recommendations though.

> if possible use .fini sections ("destructors" in GCC parlance) instead of atexit()

Why? What is the advantage of doing this? As this is non-standard we will still have to use atexit() for platforms that don't have it...and then you've got the problem of detecting which platforms have it and which don't.

> where possible use dladdr() with the address of a static variable to get the absolute path of the current object, then dlopen() that

Ok. That seems very sensible. The question though is how to do this portably? AFAICT dladdr() is a non-standard function. Most platforms do seem to have it (that I checked). But I'm sure we're going to encounter ones that don't. How do I reliably detect whether we've got it or not? Also, keep in mind that platforms where we really need this are ones where atexit() handlers don't get called on dlclose()...which seems to be unusual or very old platforms. If they're unusual or very old then they're probably more likely to not have dladdr :-(

> A third comment is that this should probably not be disablable.

That seems a strange recommendation. In the current incarnation it is not only disablable, but is disabled by default. On all the platforms that I checked this workaround is not necessary because atexit() handlers get called on dlclose(). Actually the only report we have where this is a problem is on Windows (which the move to _onexit() should fix) - so its not entirely clear whether there *are* any platforms (that are currently in use for 1.1.0) where this is needed. This whole thing is written on the assumption that somewhere at some point we are going to run into one. But why saddle the majority of implementations with this when we don't need to? (It would not be an unreasonable approach to just implement the _onexit() bit and forget about everything else until it is confirmed as a real issue)

> Also, on Windows perhaps the right thing is to use DllMain()...

I assume you mean to attach some code to DLL_PROCESS_DETACH. I'm not sure why this is better. From the documentation for _onexit():

    In the case when _onexit is called from within a DLL, routines registered with _onexit
    run on a DLL's unloading after DllMain is called with DLL_PROCESS_DETACH.

So, in the case of DLLs, calling _onexit() or doing something on DLL_PROCESS_DETACH should have the same effect. The advantage of _onexit() is that it also works where we are statically linked. If we put something explicitly in DllMain() then we also need to detect when we are being used statically and have an atexit() handler for that *as well* - which seems unnecessarily complicated?

nicowilliams commented:
.fini sections are actually part of the ELF standard, IIRC.  The nice thing about them is that they run when the object is unloaded, whereas atexit()/onexit() handlers run when the process exits (though this may vary by libc, which makes these functions rather _non-standard_, even if atexit() is in POSIX.

As to dladdr(), yes, it's not universally available.  However, surely that's already true of other C library functions used by OpenSSL, so you must have the infrastructure to detect/configure support for it.

As to disablability, if this works then why not enable by default and keep it enabled?  ISTM that OpenSSL should not get unloaded.  Perhaps I'm being pessimistic and a period of learning whether it's always better for OpenSSL to not be unloaded would be best.

mattcaswell commented:
> .fini sections are actually part of the ELF standard, IIRC. The nice thing about them is that they run when the object is unloaded, whereas atexit()/onexit() handlers run when the process exits (though this may vary by libc, which makes these functions rather non-standard, even if atexit() is in POSIX.

That makes it rather difficult to use. atexit() does run when the DSO is unloaded on most platforms...its just not standard behaviour. I'd prefer to stick with atexit() for now.

> As to dladdr(), yes, it's not universally available. However, surely that's already true of other C library functions used by OpenSSL, so you must have the infrastructure to detect/configure support for it.

Unfortunately not - we don't have any "autoconf" style capability. We use preprocessor macros to detect feature presence....which is fine if you know what your looking for :-(

nicowilliams commented:
Another thing is to use `-z nodelete` on the link-editor innovation that produces each OpenSSL object.  This marks the object as not-to-be-unloaded.  Of course you have to know whether the link-editor you're using supports this, and there's no guarantee that the RTLD will honor it (since `DF1_NODELETE` isn't critical AFAIK.

As to `.fini` vs `atexit()`, i disagree.  Both are easy to use, but the former more reliablly have the desired semantics (i.e., always runs just prior to unloading).  I mentioned `DllMain()` earlier as the WIN32 analog of `.fini` second, though i agree that `_onexit()` is mostly equivalent, except that `_onexit()` is allowed to _fail_, whereas `DllMain()` won't fail to get called.  Note too that WIN32 has an `atexit()` that does the right thing (though, again, it can fail).

nicowilliams commented:
Apologies for typos; I'm on a mobile device today.

nicowilliams commented:
For `dladdr()`, you can configure it on a poet platform basis, so do that.

nicowilliams commented:
Lastly, if you have `.fini` then you don't need `dladdr()`, and the latest is available on a strict subset of platforms that have the former, so maybe forget `dladdr()` unless you want to make sure the object does not get unloaded even if `DF1_NODELETE` could not be set or was not honored.

vdukhovni commented:
Bottom line:
* the current dlopen() of a hard-coded name is wrong.
* It is best to avoid getting unloaded, this is much more robust.
* There are multiple ways of avoiding getting unload, we should be willing to try them all
* Optimally ``-z nodelete``, this avoids the problem at .so creation time.
* Secondly ``dl_addr()`` enabled in platform recipes where we know it is present, or the user building OpenSSL promises that it is.
* Thirdly, ``.init`` and ``.fini`` sections where we can't avoid being unloaded and must fall back on  finalization.  Or the corresponding Windows features via DllMain().
* Next, only where it is known safe atexit()/onexit().
* Finally, just leak if someone is fool enough to load and unload the library repeatedly.  The leak is safer than crashing.

nicowilliams commented:
So `-z nodelete` is the right thing too do on all non-Windows systems with `ld(1)`, with a `.fini` just in case on ELF platforms/configurations and `atexit()` just in case on non-ELF platforms/configurations.  The`dladdr()` thing is still advisable to ensure `nodelete` behavior, but i understand the desire to avoid discovery/configuration of it.

@vdukhovni says that `-z nodelete` support can be discovered in the configuration (even though it's not meant to auto-discover platform features in general).

richsalz commented:
What is the current plan:  make unload be a no-op/error on linux but work on windows?  We can't make unload not work on Windows; that community expects it.  What's a cross-platform product going to have to do?

nicowilliams commented:
I would, however, abuse staying far away from `.init` sections!  Use `pthread_once()` (or the `ExecuteOnce*` interface on Windows) instead.  Historically there have been issues with `.init` section execution ordering.  Mind you, OpenSSL shouldn't be patty of any circular dependencies, so `.init` and `.fini` section execution order should not be particularly problematic for OpenSSL.  But if ever a C library wanted to depend on OpenSSL (!), then this would be a problem.

nicowilliams commented:
s/abuse/advise/

nicowilliams commented:
@richsalz I'd prefer that OpenSSL not be unloadable, thus tear down either never happen, or only at process exit.

richsalz commented:
I think it is too late for that on windows.

nicowilliams commented:
@richsalz Why is no-unload not ok on Windows?  Does it break things or would merely surprise people?

nicowilliams commented:
Oh, and the Windows equivalent if `dladdr()` is `GetModuleHandleEx()`, FYI.

vdukhovni commented:
@richsalz, 
Presumably unload will not fail, but the library will stay loaded,
so that the next module that gets loaded and needs OpenSSL will
use the already loaded version.  So it is far from clear why
anyone should care whether the unload really happened.

Indeed not unloading avoids some memory leaks, because not
all the global data allocated by OpenSSL is created via
the "once" interfaces or freed via the unload interfaces.

richsalz commented:
Time to take this discussion to openssl-dev.  As I understand things on windows, each DLL has its own malloc heap.  Prog A using L1 and L2 libs that each load openssl might end up surprised.  Maybe not.

nicowilliams commented:
Each DLL gets its own heap, but not unloading the DLL on last `FreeLibrary()` doesn't imply unbounded leaks on Windows any more than on Unix-ish systems. And you can still clean up at true process exit tinge  where suppressing false positives in valgrind/DrMemory/etcetera is important.

mattcaswell commented:
Picking comments from various people:

> As to .fini vs atexit(), i disagree. Both are easy to use, but the former more reliablly have the desired semantics (i.e., always runs just prior to unloading).

.fini doesn't sound easy at all to me...it sounds like an incredible pain. IIUC it makes use of compiler specific semantics to implement them. That is invariably going to mean that we can implement them easily for the "easy" platforms, and not at all for the "hard" platforms. But the "easy" platforms don't have the problem....they have an atexit() that works. It's the "hard" ones we need to deal with (i.e. the really old, or less common).

I am far from convinced that .fini is the way to go.

> It is best to avoid getting unloaded, this is much more robust.

I don't buy this argument either. Not getting unloaded comes with a cost...we are consuming resources for the duration of the process lifetime. If a process only needed OpenSSL once during startup and then not again then not being unloaded may be unacceptable. It's fine as a fall back where we can't do anything else. I think we should be avoiding doing it by default.

> So -z nodelete is the right thing too do on all non-Windows systems with ld(1)

Given that I don't buy the argument that we should avoid getting unloaded, I obviously don't agree with this either. Again, it would be ok as a fall back position. I don't see why we should be doing it by default. This also goes back to the "hard" platforms thing again. This is quite likely not going to be the solution on the "hard" platforms. The platforms that have nodelete are most likely the ones that have a working atexit().


For Windows I am strongly of the view that _onexit() is the way to go.

nicowilliams commented:
`.fini` sections are not a pain to use.  You write a function of no arguments returning void then use a macro to create the section calling it.  The ugliness is in a header file that has `#ifdef`s to support the various compilers.  You'd also have an `#ifdef` in initialization code to call `atexit()` (and to grab and leak a dl handle to the object you're executing in) where`'fini` sections are not supported.

The allocations to be freed at object unloading tone should be O(1) and not security sensitive ( except maybe PRNG state on some platforms? can you tell if such platforms have a proper `atexit()`?), so "leaking" them by default is tolerable.  This PR is ask about doing just that (unless you detect/configure then `#ifdef` leaking the object handle).

mattcaswell commented:
So, I think I have a possible solution to the `dladdr()` portability problem. It seems we already had some functionality to (portably) do that already in the library, however it got removed from 1.1.0 as part of a cleanup because it wasn't actually being used by anything. It exists in 1.0.2 though. I've just resurrected that bit of the code.

New commits added to this. I've also addressed ensuring the libssl does not get prematurely unloaded (previously I'd only done libcrypto).

WRT .fini and nodelete, I think those are tools in our toolbox that we can pull out if and when we need to. At the moment I think they are solving a problem that we don't yet have (and might never have). If we encounter platforms which have problems we can look at the specifics of those platforms and decide the right way ahead.

I would still like to implement a test for this, but I've not got to that yet.

t-j-h commented:
BTW I do think need a test case covering this area ... so we know if we are hitting a platform where the assumptions are not valid. However as we know this is specifically broken on Win32 and it is a primary platform I provided the "approved" response.

mattcaswell commented:
Thanks @t-j-h for the review, but the approval was a bit premature (this was still marked as WIP). I have now reworked this quite a bit on the basis of experience implementing a test.

It turns out that the original approach just didn't work. Calling dlclose() from within an atexit() handler doesn't seem to work too well and was causing crashes. This left me with no real choice but to go with @nicowilliams original recommendation of leaving the libraries permanently loaded by default. Given that approach, my objection to nodelete goes away too, so I have implemented that (for Linux) as well.

I have now had the opportunity to test this on a few different platforms and it seems to hold up, so I'm now removing the WIP label. @t-j-h can you take another look?

jay commented:
> The Windows implementation of atexit() does not get called on FreeLibrary() (the equivalent of dlclose()).

As discussed in curl/curl#1055 it is documented in [runtime library behavior](https://msdn.microsoft.com/en-us/library/988ye33t.aspx) that a DLL's atexit handlers are called on process detach (eg FreeLibrary 0 refs left). But that's for the _Microsoft_ CRT. I wrote a [test](https://gist.github.com/jay/423e12ea7f5b86c72f59719c8f5a006c) to demonstrate it. It is interesting that it is not documented in their [atexit](https://msdn.microsoft.com/en-us/library/tze57ck3.aspx) pages.

The problem from what I could gather appears to be CRT alternatives, not Microsoft's. Reportedly in the case of Embarcadero CRT it is a problem (did not test). But I did test if a program that uses Microsoft's CRT loads a DLL with the cygwin CRT, and that is also a problem because the DLL's atexit handler isn't called. Cygwin by itself though is not a problem.

This is not an objection of your switch to _onexit, however depending how you coded it that could cause a build issue if alternative CRTs don't have it.

mattcaswell commented:
@jay, the swap to `_onexit()` only happens if `_WIN32_WINNT` is defined. Does that happen with other CRTs? The fix in this MR should work even where `atexit()` is used though.

mattcaswell commented:
Ping @t-j-h.

jay commented:
> the swap to _onexit() only happens if _WIN32_WINNT is defined. Does that happen with other CRTs?

Yes it's possible. That define is not specific to Microsoft's CRT, it's basically used to hide/expose API depending on what's targeted. Windows API may be used at the same time as an alternative CRT. Someone could define `_WIN32_WINNT` when building your project or your project could define it or if your source includes Windows include (or something pulls it in) then it will define it if not already set (_assuming your Windows includes are >=2008 and support `_WIN32_WINNT` and not just `WINVER`_).

AFAIK `_onexit` is Microsoft CRT specific. You could probably do `#if defined(_MSC_VER) || defined(_ONEXIT_T_DEFINED) || defined(__MINGW32__)`

mattcaswell commented:
I've actually decided to drop the `_onexit` change altogether. It's not actually needed now that OpenSSL will remain loaded until exit anyway. This way we don't have to worry about CRT versions etc.

Ping @t-j-h.

t-j-h commented:
It seems to me that we are doing a lot more work to achieve the outcome that is necessary - in a Win32 environment at least. Basically we want to increment the reference count so until the process exists we have a reference to the loaded OpenSSL code.

Without thinking too hard about it I thought that GetModuleHandleEx had the required functionality in it. I'm always mindful of hitting toolhelp32 ... 

https://msdn.microsoft.com/en-us/library/windows/desktop/ms683200(v=vs.85).aspx

@mattcaswell can you take a look at that and see if it looks like an alternative path - what you've got in place looks workable to me but I think checking to see if there is a faster (less code, less overhead) solution in a Win32 environment is worthwhile.

jay commented:
We had considered something similar in the curl project, see curl/curl#997. It's on the backburner for now but basically it looks like this (untested)
~~~c
++    /* Get a handle to the program module */
++    program = GetModuleHandle(NULL);
++    if(!program)
++      return CURLE_FAILED_INIT;
++
++    /* Get a handle to the module containing this code and increase its
++       refcount */
++    ret = GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS,
++                            (void *)function_in_our_dll, &libcurl);
++    if(!ret)
++      return CURLE_FAILED_INIT;
++
++    /* If the module handles differ then our module is a DLL. */
++    if(program != libcurl)
++      dll = libcurl;
~~~
However I was bothered by this because we have users who compile libcurl as a static library and then include it in their DLL, so in that case we are not just keeping libcurl loaded we are keeping their DLL loaded which is unexpected. The alternative though is a possible crash after the DLL is unloaded if any libcurl started threads are running. For now it is just documented as 'CAUTION' in [curl_global_cleanup](https://curl.haxx.se/libcurl/c/curl_global_cleanup.html).

mattcaswell commented:
@t-j-h, the win32_pathbyaddr stuff which I think is what you are referring to, was just a resurrection of code that already existed in 1.0.2. All I did was reintroduce it. Looking at GetModuleHandleEx() it does seem to do the required job....actually even better it has a GET_MODULE_HANDLE_EX_FLAG_PIN flag, which basically does the same think as nodelete. So I have refactored things to remove the win32_pathbyaddr() code, and added a special case for windows to use GetModuleHandleEx() in the init code. I've also rebased to remove some merge conflicts that github was reporting.

vdukhovni commented:
I see new code to handle libcrypto deinit, but nothing for libssl.  Shouldn't there be similar changes in ssl_init.c?

mattcaswell commented:
@vdukhovni, the libcrypto changes also handle keeping libssl in memory until exit. libssl does not register an `atexit()` handler directly. Instead it calls `OPENSSL_atexit()` and provides a pointer to its handler. The `OPENSSL_atexit()` implementation has changed in this PR to ensure that the module containing the handler (libssl) is kept in memory until process exit.

mattcaswell commented:
Ping @t-j-h

mattcaswell commented:
Grrr...I spotted a problem while cherry-picking this. The symbol versions for the DSO functions in libcrypto.num are wrong for backport to 1.1.0. They need to be for 1.1.0c. Fixed now.

@t-j-h can you reconfirm?

t-j-h commented:
+1 reconfirmed

mattcaswell commented:
Pushed. Thanks.
